packages-exp/app-check-exp/src/internal-api.ts

```diff
@@ -0,0 +1,311 @@
+// Format left open for possible dynamic error values and other fields in the future.
+export const defaultTokenErrorData = { error: 'UNKNOWN_ERROR' };
+
+/**
+ * Stringify and base64 encode token error data.
+ *
+ * @param tokenError Error data, currently hardcoded.
+ */
+export function formatDummyToken(
+  tokenErrorData: Record<string, string>
+): string {
+  return base64.encodeString(
+    JSON.stringify(tokenErrorData),
+    /* webSafe= */ false
+  );
+}
+
+/**
+ * This function will always resolve.
+ * The result will contain an error field if there is any error.
+ * In case there is an error, the token field in the result will be populated with a dummy value
+ */
+export async function getToken(
+  app: FirebaseApp,
+  platformLoggerProvider: Provider<'platform-logger'>,
+  forceRefresh = false
+): Promise<AppCheckTokenResult> {
+  ensureActivated(app);
+  /**
+   * DEBUG MODE
+   * return the debug token directly
+   */
+  if (isDebugMode()) {
+    const tokenFromDebugExchange: AppCheckTokenInternal = await exchangeToken(
+      getExchangeDebugTokenRequest(app, await getDebugToken()),
+      platformLoggerProvider
+    );
+    return { token: tokenFromDebugExchange.token };
+  }
+
+  const state = getState(app);
+
+  let token: AppCheckTokenInternal | undefined = state.token;
+  let error: Error | undefined = undefined;
+
+  /**
+   * try to load token from indexedDB if it's the first time this function is called
+   */
+  if (!token) {
+    // readTokenFromStorage() always resolves. In case of an error, it resolves with `undefined`.
+    const cachedToken = await readTokenFromStorage(app);
+    if (cachedToken && isValid(cachedToken)) {
+      token = cachedToken;
+
+      setState(app, { ...state, token });
+      // notify all listeners with the cached token
+      notifyTokenListeners(app, { token: token.token });
+    }
+  }
+
+  // return the cached token if it's valid
+  if (!forceRefresh && token && isValid(token)) {
+    return {
+      token: token.token
+    };
+  }
+
+  /**
+   * request a new token
+   */
+  try {
+    if (state.provider instanceof ReCaptchaV3ProviderImpl) {
+      token = await state.provider.getToken();
+    } else if (state.provider) {
+      // custom provider
+      const customToken = await state.provider.getToken();
+      // Try to extract IAT from custom token, in case this token is not
+      // being newly issued. JWT timestamps are in seconds since epoch.
+      const issuedAtTimeSeconds = issuedAtTime(customToken.token);
+      // Very basic validation, use current timestamp as IAT if JWT
+      // has no `iat` field or value is out of bounds.
+      const issuedAtTimeMillis =
+        issuedAtTimeSeconds !== null &&
+        issuedAtTimeSeconds < Date.now() &&
+        issuedAtTimeSeconds > 0
+          ? issuedAtTimeSeconds * 1000
+          : Date.now();
+
+      token = { ...customToken, issuedAtTimeMillis };
+    }
```

Comment:
We need to encapsulate the special handling for custom tokens in the customProvider, so we don't need to do instanceof check which breaks tree shaking.

We will need to wrap the user provided customProvider in our own class which can do the special handling. We probably need to ask developers to create a customProvider using a class, or a function we provide, something like:

```
import { initializeAppCheck, CustomProvider } from 'firebase/app-check';

initializeAppCheck(app, new CustomProvider({
    getToken: () => {}
    ...
}));

```

Reply:
Done.